File: packages/provide-uterm-go/detection/extractor_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package detection

import (
	"strings"
	"testing"
)

func cfg(field, ftype, regex string) map[string]any {
	return map[string]any{"field": field, "type": ftype, "regex": regex}
}

func TestExtractStringField(t *testing.T) {
	result := ExtractKVWithValidation("Player: TestUser\nScore: 1000", cfg("player", "string", `Player:\s*(\w+)`), false)
	if result == nil || result["player"] != "TestUser" {
		t.Fatalf("result = %v", result)
	}
}

func TestExtractIntField(t *testing.T) {
	result := ExtractKVWithValidation("Score: 1000\nLevel: 5", cfg("score", "int", `Score:\s*(\d+)`), false)
	if result == nil || result["score"] != 1000 {
		t.Fatalf("result = %v", result)
	}
}

func TestExtractIntWithCommas(t *testing.T) {
	result := ExtractKVWithValidation("Credits: 1,234,567", cfg("credits", "int", `Credits:\s*([\d,]+)`), false)
	if result == nil || result["credits"] != 1234567 {
		t.Fatalf("result = %v", result)
	}
}

func TestExtractFloatField(t *testing.T) {
	result := ExtractKVWithValidation("Temperature: 98.6 degrees", cfg("temp", "float", `Temperature:\s*([\d.]+)`), false)
	if result == nil || result["temp"] != 98.6 {
		t.Fatalf("result = %v", result)
	}
	result = ExtractKVWithValidation("Value: 1,234.56", cfg("value", "float", `Value:\s*([\d,.]+)`), false)
	if result == nil || result["value"] != 1234.56 {
		t.Fatalf("float commas = %v", result)
	}
}

func TestExtractBoolField(t *testing.T) {
	for _, v := range []string{"true", "True", "TRUE", "yes", "Yes", "y", "Y", "1", "on", "ON"} {
		result := ExtractKVWithValidation("Flag: "+v, cfg("flag", "bool", `Flag:\s*(\w+)`), false)
		if result == nil || result["flag"] != true {
			t.Errorf("true variant %q -> %v", v, result)
		}
	}
	for _, v := range []string{"false", "False", "FALSE", "no", "No", "n", "N", "0", "off", "OFF"} {
		result := ExtractKVWithValidation("Flag: "+v, cfg("flag", "bool", `Flag:\s*(\w+)`), false)
		if result == nil || result["flag"] != false {
			t.Errorf("false variant %q -> %v", v, result)
		}
	}
	if ExtractKVWithValidation("Flag: maybe", cfg("flag", "bool", `Flag:\s*(\w+)`), false) != nil {
		t.Error("invalid bool should fail extraction")
	}
}

func TestExtractConversionFailures(t *testing.T) {
	if ExtractKVWithValidation("Score: abc", cfg("score", "int", `Score:\s*(\w+)`), false) != nil {
		t.Error("int conversion failure")
	}
	if ExtractKVWithValidation("Temp: abc", cfg("temp", "float", `Temp:\s*(\w+)`), false) != nil {
		t.Error("float conversion failure")
	}
}

func TestExtractUnknownTypeReturnsString(t *testing.T) {
	result := ExtractKVWithValidation("Value: test123", cfg("value", "unknown", `Value:\s*(\w+)`), false)
	if result == nil || result["value"] != "test123" {
		t.Fatalf("result = %v", result)
	}
}

func TestExtractNoMatchReturnsNil(t *testing.T) {
	if ExtractKVWithValidation("Player: TestUser", cfg("score", "int", `Score:\s*(\d+)`), false) != nil {
		t.Error("no match")
	}
}

func TestExtractInvalidConfig(t *testing.T) {
	if ExtractKVWithValidation("x", nil, false) != nil {
		t.Error("nil config")
	}
	if ExtractKVWithValidation("x", map[string]any{}, false) != nil {
		t.Error("empty dict config")
	}
	// dict without "field" key
	if ExtractKVWithValidation("Score: 100", map[string]any{"regex": `Score:\s*(\d+)`}, false) != nil {
		t.Error("dict without field")
	}
	// unsupported shape
	if ExtractKVWithValidation("x", 42, false) != nil {
		t.Error("int config")
	}
	// empty field / empty pattern entries skipped
	configs := []any{
		map[string]any{"field": "", "regex": `Score:\s*(\d+)`, "type": "int"},
		map[string]any{"field": "score", "regex": "", "type": "int"},
	}
	if ExtractKVWithValidation("Score: 100", configs, false) != nil {
		t.Error("skipped entries")
	}
	// invalid regex entry skipped
	if ExtractKVWithValidation("Score: 100", cfg("score", "int", `[bad`), false) != nil {
		t.Error("invalid regex")
	}
	// non-map entries in []any ignored
	if got := ExtractKVWithValidation("Score: 100", []any{"junk", cfg("score", "int", `Score:\s*(\d+)`)}, false); got == nil || got["score"] != 100 {
		t.Errorf("mixed list = %v", got)
	}
}

func TestExtractMultipleFields(t *testing.T) {
	screen := "\nPlayer: TestUser\nScore: 1000\nLevel: 5\n"
	configs := []any{
		cfg("player", "string", `Player:\s*(\w+)`),
		cfg("score", "int", `Score:\s*(\d+)`),
		cfg("level", "int", `Level:\s*(\d+)`),
	}
	result := ExtractKVWithValidation(screen, configs, false)
	if result == nil || result["player"] != "TestUser" || result["score"] != 1000 || result["level"] != 5 {
		t.Fatalf("result = %v", result)
	}
	// typed-slice form
	typed := []map[string]any{cfg("player", "string", `Player:\s*(\w+)`)}
	if got := ExtractKVWithValidation(screen, typed, false); got == nil || got["player"] != "TestUser" {
		t.Errorf("typed slice = %v", got)
	}
}

func TestExtractPartialMatch(t *testing.T) {
	configs := []any{
		cfg("player", "string", `Player:\s*(\w+)`),
		cfg("level", "int", `Level:\s*(\d+)`),
	}
	result := ExtractKVWithValidation("Player: TestUser\nScore: 1000", configs, false)
	if result == nil || result["player"] != "TestUser" {
		t.Fatal("partial")
	}
	if _, has := result["level"]; has {
		t.Error("level should be absent")
	}
}

func TestExtractLastMatchWins(t *testing.T) {
	// scroll history: most recent value at the end
	result := ExtractKVWithValidation("Sector 1\nSector 42", cfg("sector", "int", `Sector\s+(\d+)`), false)
	if result == nil || result["sector"] != 42 {
		t.Fatalf("last match = %v", result)
	}
}

func TestExtractNoCaptureGroupUsesWholeMatch(t *testing.T) {
	result := ExtractKVWithValidation("Score: 1000", cfg("score", "int", `\d+`), false)
	if result == nil || result["score"] != 1000 {
		t.Fatalf("whole match = %v", result)
	}
	strRes := ExtractKVWithValidation("Hello World", cfg("match", "string", `Hello`), false)
	if strRes == nil || strRes["match"] != "Hello" {
		t.Fatalf("string whole match = %v", strRes)
	}
}

func TestExtractFirstGroupOfMany(t *testing.T) {
	result := ExtractKVWithValidation("Player: TestUser (Level 5)",
		cfg("player", "string", `Player:\s*(\w+)\s*\(Level\s*(\d+)\)`), false)
	if result == nil || result["player"] != "TestUser" {
		t.Fatalf("first group = %v", result)
	}
}

func TestExtractOptionalGroupNotParticipating(t *testing.T) {
	// group 1 exists but does not participate -> Python lastindex is None,
	// falls back to whole match.
	result := ExtractKVWithValidation("xyz", cfg("v", "string", `xyz(\d+)?`), false)
	if result == nil || result["v"] != "xyz" {
		t.Fatalf("optional group = %v", result)
	}
}

func TestExtractCaseInsensitiveMultiline(t *testing.T) {
	result := ExtractKVWithValidation("PLAYER: TestUser", cfg("player", "string", `player:\s*(\w+)`), false)
	if result == nil || result["player"] != "TestUser" {
		t.Fatalf("case-insensitive = %v", result)
	}
}

// --- validation --------------------------------------------------------------

func vmap(res map[string]any, t *testing.T) map[string]any {
	t.Helper()
	v, ok := res["_validation"].(map[string]any)
	if !ok {
		t.Fatalf("no _validation in %v", res)
	}
	return v
}

func verrs(res map[string]any, t *testing.T) []string {
	t.Helper()
	return vmap(res, t)["errors"].([]string)
}

func TestValidationPasses(t *testing.T) {
	c := cfg("score", "int", `Score:\s*(\d+)`)
	c["validate"] = map[string]any{"min": 0, "max": 2000}
	result := ExtractKV("Score: 1000", c)
	if result == nil || vmap(result, t)["valid"] != true || len(verrs(result, t)) != 0 {
		t.Fatalf("result = %v", result)
	}
}

func TestValidationMinMax(t *testing.T) {
	c := cfg("score", "int", `Score:\s*(-?\d+)`)
	c["validate"] = map[string]any{"min": 0}
	result := ExtractKV("Score: -10", c)
	if vmap(result, t)["valid"] != false || !containsSubstr(verrs(result, t), "below min") {
		t.Errorf("min: %v", result)
	}
	c2 := cfg("score", "int", `Score:\s*(\d+)`)
	c2["validate"] = map[string]any{"max": 2000}
	result2 := ExtractKV("Score: 5000", c2)
	if vmap(result2, t)["valid"] != false || !containsSubstr(verrs(result2, t), "exceeds max") {
		t.Errorf("max: %v", result2)
	}
}

func TestValidationRequiredMissing(t *testing.T) {
	c := cfg("score", "int", `Score:\s*(\d+)`)
	c["required"] = true
	// Nothing extracted at all -> nil result (matches Python)
	if ExtractKV("Player: TestUser", c) != nil {
		t.Error("nothing extracted -> nil")
	}
	// present
	result := ExtractKV("Score: 1000", c)
	if result == nil || result["score"] != 1000 || vmap(result, t)["valid"] != true {
		t.Errorf("present = %v", result)
	}
	// required missing among other extracted fields
	configs := []any{
		cfg("score", "int", `Score:\s*(\d+)`),
		func() map[string]any { c := cfg("level", "int", `Level:\s*(\d+)`); c["required"] = true; return c }(),
	}
	res := ExtractKV("Score: 100", configs)
	if vmap(res, t)["valid"] != false || !containsSubstr(verrs(res, t), "required but not found") {
		t.Errorf("required among = %v", res)
	}
}

func TestValidationPattern(t *testing.T) {
	c := cfg("name", "string", `Name:\s*(\S+)`)
	c["validate"] = map[string]any{"pattern": `^[A-Za-z]+$`}
	result := ExtractKV("Name: Test123", c)
	if vmap(result, t)["valid"] != false || !containsSubstr(verrs(result, t), "does not match pattern") {
		t.Errorf("pattern fail = %v", result)
	}
	pass := ExtractKV("Name: Alice", c)
	if vmap(pass, t)["valid"] != true {
		t.Errorf("pattern pass = %v", pass)
	}
	// invalid validation pattern regex -> counted as mismatch
	c2 := cfg("name", "string", `Name:\s*(\S+)`)
	c2["validate"] = map[string]any{"pattern": `[bad`}
	bad := ExtractKV("Name: Alice", c2)
	if vmap(bad, t)["valid"] != false {
		t.Errorf("bad pattern = %v", bad)
	}
	// re.match semantics: pattern must match at position 0
	c3 := cfg("name", "string", `Name:\s*(\S+)`)
	c3["validate"] = map[string]any{"pattern": `lice`}
	mid := ExtractKV("Name: Alice", c3)
	if vmap(mid, t)["valid"] != false {
		t.Errorf("mid-string match must fail re.match semantics: %v", mid)
	}
}

func TestValidationAllowedValues(t *testing.T) {
	c := cfg("mode", "string", `Mode:\s*(\w+)`)
	c["validate"] = map[string]any{"allowed_values": []any{"Normal", "Test", "Production"}}
	result := ExtractKV("Mode: Debug", c)
	if vmap(result, t)["valid"] != false || !containsSubstr(verrs(result, t), "not in allowed values") {
		t.Errorf("allowed fail = %v", result)
	}
	pass := ExtractKV("Mode: Normal", c)
	if vmap(pass, t)["valid"] != true {
		t.Errorf("allowed pass = %v", pass)
	}
	// []string form
	c2 := cfg("mode", "string", `Mode:\s*(\w+)`)
	c2["validate"] = map[string]any{"allowed_values": []string{"Normal"}}
	if vmap(ExtractKV("Mode: Normal", c2), t)["valid"] != true {
		t.Error("[]string allowed")
	}
	// unsupported allowed shape -> not allowed
	c3 := cfg("mode", "string", `Mode:\s*(\w+)`)
	c3["validate"] = map[string]any{"allowed_values": "Normal"}
	if vmap(ExtractKV("Mode: Normal", c3), t)["valid"] != false {
		t.Error("string allowed shape rejected")
	}
}

func TestValidationMultipleFields(t *testing.T) {
	configs := []any{
		func() map[string]any {
			c := cfg("score", "int", `Score:\s*(\d+)`)
			c["validate"] = map[string]any{"min": 0, "max": 2000}
			return c
		}(),
		func() map[string]any {
			c := cfg("level", "int", `Level:\s*(\d+)`)
			c["validate"] = map[string]any{"min": 1, "max": 100}
			return c
		}(),
	}
	result := ExtractKV("Score: 1000\nLevel: 150", configs)
	if result == nil || result["score"] != 1000 || result["level"] != 150 {
		t.Fatalf("values = %v", result)
	}
	if vmap(result, t)["valid"] != false {
		t.Error("level > 100 must fail")
	}
}

func TestValidateInternals(t *testing.T) {
	// non-string field key skipped
	res := validateExtracted(map[string]any{"score": 100}, []map[string]any{{"field": 42}})
	if res["valid"] != true {
		t.Error("non-string field skipped")
	}
	// float type check fails on string value
	res = validateExtracted(map[string]any{"value": "not_a_float"}, []map[string]any{{"field": "value", "type": "float"}})
	if res["valid"] != true && !containsSubstr(res["errors"].([]string), "expected float") {
		t.Errorf("float type = %v", res)
	}
	// float min/max
	res = validateExtracted(map[string]any{"temp": 200.0},
		[]map[string]any{{"field": "temp", "type": "float", "validate": map[string]any{"min": 0.0, "max": 100.0}}})
	if res["valid"] != false || !containsSubstr(res["errors"].([]string), "exceeds max") {
		t.Errorf("float max = %v", res)
	}
	res = validateExtracted(map[string]any{"temp": -5.0},
		[]map[string]any{{"field": "temp", "type": "float", "validate": map[string]any{"min": 0.0}}})
	if res["valid"] != false || !containsSubstr(res["errors"].([]string), "below min") {
		t.Errorf("float min = %v", res)
	}
	// string type check fails on int value
	res = validateExtracted(map[string]any{"name": 42}, []map[string]any{{"field": "name", "type": "string"}})
	if res["valid"] != false || !containsSubstr(res["errors"].([]string), "expected string") {
		t.Errorf("string type = %v", res)
	}
	// int type check fails on string value
	res = validateExtracted(map[string]any{"score": "not_an_int"}, []map[string]any{{"field": "score", "type": "int"}})
	if res["valid"] != false || !containsSubstr(res["errors"].([]string), "expected int") {
		t.Errorf("int type = %v", res)
	}
	// value absent and not required -> skipped
	res = validateExtracted(map[string]any{},
		[]map[string]any{{"field": "score", "type": "int", "required": false, "validate": map[string]any{"min": 0}}})
	if res["valid"] != true {
		t.Error("absent not-required skipped")
	}
	// unknown type falls through
	res = validateExtracted(map[string]any{"flag": true}, []map[string]any{{"field": "flag", "type": "bool"}})
	if res["valid"] != true {
		t.Error("bool type no checks")
	}
	// int64 value passes int check
	res = validateExtracted(map[string]any{"n": int64(5)},
		[]map[string]any{{"field": "n", "type": "int", "validate": map[string]any{"min": 1}}})
	if res["valid"] != true {
		t.Errorf("int64 = %v", res)
	}
}

func TestPyFormattersAndHelpers(t *testing.T) {
	if pyTypeName("x") != "str" || pyTypeName(1) != "int" || pyTypeName(int64(1)) != "int" ||
		pyTypeName(1.5) != "float" || pyTypeName(true) != "bool" || pyTypeName(nil) != "NoneType" {
		t.Error("pyTypeName")
	}
	if pyNum(5) != "5" || pyNum(int64(7)) != "7" || pyNum(2.5) != "2.5" || pyNum("x") != "x" {
		t.Error("pyNum")
	}
	if pyList([]any{"a", 1}) != "['a', 1]" {
		t.Errorf("pyList = %q", pyList([]any{"a", 1}))
	}
	if pyList("notalist") != "notalist" {
		t.Errorf("pyList passthrough = %q", pyList("notalist"))
	}
	// min/max rules with unparseable bounds are ignored
	res := validateExtracted(map[string]any{"n": 5},
		[]map[string]any{{"field": "n", "type": "int", "validate": map[string]any{"min": "x", "max": "y"}}})
	if res["valid"] != true {
		t.Errorf("string bounds ignored = %v", res)
	}
}

func containsSubstr(errs []string, sub string) bool {
	for _, e := range errs {
		if strings.Contains(e, sub) {
			return true
		}
	}
	return false
}

File: packages/provide-uterm-ts/src/detection/extractor.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { describe, expect, it } from "vitest";
import { loadGolden, must } from "../testing/golden.ts";
import { convertType, type ExtractConfig, extractKV } from "./index.ts";

interface ConvertCase {
  name: string;
  value: string;
  type: string;
  ok: boolean;
  value_out?: unknown;
  is_nan?: boolean;
  is_infinite?: boolean;
  is_bool?: boolean;
  error?: string;
}

interface ExtractorGolden {
  screen: string;
  single: Array<{ name: string; screen: string; config: ExtractConfig; result: Record<string, unknown> | null }>;
  convert: Array<ConvertCase & { value: string; ok: boolean }>;
  extract: Array<{
    name: string;
    screen: string;
    config: ExtractConfig | ExtractConfig[] | null;
    result: Record<string, unknown> | null;
  }>;
  unvalidated: Array<{ name: string; result: Record<string, unknown> | null }>;
  convenience_matches_extract: boolean;
}

const golden = loadGolden<ExtractorGolden>("extractor_golden.json");

/**
 * Rebuild what JSON could not carry.
 *
 * A float field can extract to an infinity or a NaN, because Python's
 * `float()` reads those words. Neither survives JSON, so the corpus records
 * a marker.
 */
function rebuilt(result: Record<string, unknown> | null): Record<string, unknown> | null {
  if (result === null) {
    return null;
  }
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(result)) {
    const marker = (value as { __float__?: string } | null)?.__float__;
    out[key] =
      marker === "nan"
        ? Number.NaN
        : marker === "inf"
          ? Number.POSITIVE_INFINITY
          : marker === "-inf"
            ? Number.NEGATIVE_INFINITY
            : value;
  }
  return out;
}

/** The recorded conversion for a named case. */
function convertCase(name: string) {
  return golden.convert.find((entry) => entry.name === name);
}

/** The recorded extraction for a named case. */
function extractCase(name: string) {
  return golden.extract.find((entry) => entry.name === name);
}

describe("pulling one field out of a screen", () => {
  it.each(golden.single)("$name", (record) => {
    expect(extractKV(record.screen, record.config, false) ?? null).toStrictEqual(rebuilt(record.result));
  });

  it("takes the last match, not the first", () => {
    // A screen buffer holds scroll history, so the same label appears many
    // times and only the bottom one is current. Reading the first would spend
    // a credit balance from several screens ago.
    expect(extractCase("several fields")?.result?.sector).toBe(42);
    expect(golden.screen).toContain("Sector  1");
    expect(golden.single.find((entry) => entry.name === "the last of several")?.result?.sector).toBe(42);
  });

  it("prefers a capture group over the whole match", () => {
    // So a pattern written with parentheses yields the number rather than the
    // label and the number together.
    expect(golden.single.find((entry) => entry.name === "an int")?.result?.sector).toBe(42);
  });

  it("uses the whole match when there is no group", () => {
    expect(golden.single.find((entry) => entry.name === "no capture group takes the whole match")?.result?.sector).toBe(
      42,
    );
  });

  it("takes the first group when there are several", () => {
    expect(golden.single.find((entry) => entry.name === "two groups take the first")?.result?.pair).toBe("1");
  });

  it("drops a field it could not convert rather than guessing", () => {
    // A zero where the screen said "abc" is worse than nothing: a caller
    // cannot tell it apart from a real zero.
    for (const name of ["an int that is not one", "a float-shaped string as an int", "a bool that is not one"]) {
      expect(golden.single.find((entry) => entry.name === name)?.result).toBeNull();
    }
  });

  it("treats a field name of zero as no field name", () => {
    // Falsy in the reference, so the whole config is skipped rather than a
    // field called "0" appearing in the result.
    expect(extractCase("a field name that is zero")?.result).toBeNull();
  });

  it("does not turn an absent pattern into the word undefined", () => {
    // A config with no regex is skipped. Stringifying the missing value would
    // make it a pattern that matches screens saying "undefined".
    expect(extractCase("a screen that says undefined")?.result).toBeNull();
  });

  it("needs both a field name and a pattern", () => {
    for (const name of ["no field name", "no regex", "an empty field name", "an empty regex"]) {
      expect(golden.single.find((entry) => entry.name === name)?.result).toBeNull();
    }
  });

  it("matches case-insensitively and per line", () => {
    // Both are on by default. A screen's case drifts between versions of a
    // program, and a value can be anywhere on it.
    expect(golden.single.find((entry) => entry.name === "case-insensitive by default")?.result?.sector).toBe(42);
    expect(golden.single.find((entry) => entry.name === "multiline by default")?.result?.sector).toBe(42);
  });

  it("strips the whitespace around a value", () => {
    expect(golden.single.find((entry) => entry.name === "whitespace is stripped")?.result?.name).toBe("Alice");
  });

  it("falls back to text for a type it does not know", () => {
    // Better a string than nothing: the rule author still gets their value.
    expect(golden.single.find((entry) => entry.name === "an unknown type falls back to text")?.result?.sector).toBe(
      "42",
    );
  });
});

describe("converting a matched string", () => {
  it.each(golden.convert)("$name as $type", (record) => {
    if (!record.ok) {
      expect(convertType(record.value, record.type)).toBeUndefined();
      return;
    }
    const converted = convertType(record.value, record.type);
    if (record.is_nan === true) {
      expect(converted).toBeNaN();
    } else if (record.is_infinite !== undefined) {
      expect(converted).toBe(record.is_infinite ? Number.POSITIVE_INFINITY : Number.NEGATIVE_INFINITY);
    } else {
      expect(converted).toStrictEqual(record.value_out);
    }
  });

  it("strips thousands separators", () => {
    // Screens write numbers for people to read.
    expect(convertCase("an integer with separators")?.value_out).toBe(1234567);
    expect(convertCase("a float with separators")?.value_out).toBe(1234.5);
  });

  it("refuses a float-shaped string as an integer", () => {
    // Truncating it would turn 1.9 sectors into sector 1 without saying so.
    expect(convertCase("a float as an int")?.ok).toBe(false);
    expect(convertType("1.9", "int")).toBeUndefined();
  });

  it("accepts the underscores Python's own literals allow", () => {
    expect(convertCase("an integer with underscores")?.value_out).toBe(1000);
  });

  it("refuses an empty string as a number", () => {
    // Number("") is zero in this language, which is the single most likely
    // way an unread screen becomes a plausible-looking value.
    expect(convertType("", "int")).toBeUndefined();
    expect(convertType("", "float")).toBeUndefined();
    expect(convertType("   ", "float")).toBeUndefined();
  });

  it("reads the float words Python reads", () => {
    expect(convertType("inf", "float")).toBe(Number.POSITIVE_INFINITY);
    expect(convertType("-inf", "float")).toBe(Number.NEGATIVE_INFINITY);
    expect(convertType("nan", "float")).toBeNaN();
    expect(convertType("infinity", "float")).toBe(Number.POSITIVE_INFINITY);
  });

  it("does not read the ones it does not", () => {
    // A hexadecimal literal is a number to this language and not to Python,
    // so a screen showing "0x10" must not become 16.
    expect(convertType("0x10", "int")).toBeUndefined();
    expect(convertType("0x10", "float")).toBeUndefined();
    expect(convertType("1e5", "int")).toBeUndefined();
  });

  it("takes the truthy words and the falsy ones", () => {
    for (const name of ["true", "yes", "y", "one", "on", "upper case true", "mixed case yes"]) {
      expect(convertCase(name)?.value_out).toBe(true);
    }
    for (const name of ["false", "no", "n", "zero", "off"]) {
      expect(convertCase(name)?.value_out).toBe(false);
    }
  });

  it("refuses a word that is neither", () => {
    // "maybe" is not a decision, and guessing one is how an agent confirms
    // something it was not asked to confirm.
    expect(convertCase("something else as a bool")?.ok).toBe(false);
    expect(convertType("maybe", "bool")).toBeUndefined();
  });

  it("keeps a bool a bool", () => {
    // Not one and zero: a caller checking `=== true` would otherwise never
    // match.
    expect(convertType("yes", "bool")).toBe(true);
    expect(convertType("no", "bool")).toBe(false);
  });
});

describe("extracting a whole configuration", () => {
  it.each(golden.extract)("$name", (record) => {
    expect(extractKV(record.screen, record.config) ?? null).toStrictEqual(rebuilt(record.result));
  });

  it.each(golden.unvalidated)("$name, without validation", (record) => {
    const source = extractCase(record.name);
    expect(extractKV(source?.screen as string, source?.config as ExtractConfig[], false) ?? null).toStrictEqual(
      rebuilt(record.result),
    );
  });

  it("accepts a single configuration as well as a list", () => {
    expect(extractCase("a single config rather than a list")?.result?.sector).toBe(42);
  });

  it("says nothing when there is nothing to do", () => {
    for (const name of ["nothing configured", "an empty list", "an empty dict"]) {
      expect(extractCase(name)?.result).toBeNull();
    }
  });

  it("refuses a configuration that is neither a list nor a field", () => {
    for (const name of ["a config that is neither", "a dict with no field key"]) {
      expect(extractCase(name)?.result).toBeNull();
    }
  });

  it("says nothing when nothing matched", () => {
    // An empty result and no result are different: the caller learns the
    // screen was not the one they expected.
    expect(extractCase("nothing matched")?.result).toBeNull();
  });

  it("keeps the fields it found when one is missing", () => {
    // Partial data beats none — the caller can still see what was on screen.
    const result = extractCase("one field matched and one not")?.result;
    expect(result?.sector).toBe(42);
    expect(Object.hasOwn(result as object, "credits")).toBe(false);
  });

  it("reports validation rather than enforcing it", () => {
    // The values come back alongside the verdict, so a caller can act on one
    // it knows is out of range — or refuse to. Dropping it would leave them
    // unable to tell "absent" from "implausible".
    const result = extractCase("a value below its minimum")?.result;
    expect(result?.sector).toBe(3);
    expect((must(result, "the below-minimum extraction")._validation as { valid: boolean }).valid).toBe(false);
  });

  it("leaves the report out when it was not asked for", () => {
    const index = golden.extract.findIndex((entry) => entry.name === "several fields");
    expect(Object.hasOwn(golden.unvalidated[index]?.result as object, "_validation")).toBe(false);
    expect(Object.hasOwn(golden.extract[index]?.result as object, "_validation")).toBe(true);
  });
});

describe("what validation reports", () => {
  /** The verdict recorded for a named case. */
  function verdict(name: string) {
    return extractCase(name)?.result?._validation as { valid: boolean; errors: string[] } | undefined;
  }

  it("passes a value inside its range", () => {
    expect(verdict("a value inside its range")?.valid).toBe(true);
    expect(verdict("a value inside its range")?.errors).toStrictEqual([]);
  });

  it("treats the bounds as inclusive", () => {
    // A value exactly on its limit is allowed; excluding it would reject the
    // first and last legitimate values.
    expect(verdict("a value exactly on its bounds")?.valid).toBe(true);
  });

  it("names the field, the value and the bound it broke", () => {
    // The message is what an operator reads; a bare "invalid" would send them
    // back to the rules file to guess which one.
    expect(verdict("a value below its minimum")?.errors[0]).toContain("sector");
    expect(verdict("a value below its minimum")?.errors[0]).toContain("3");
    expect(verdict("a value above its maximum")?.errors[0]).toContain("100");
  });

  it("reports a required field that was not found", () => {
    const result = verdict("a required field that is missing");
    expect(result?.valid).toBe(false);
    expect(result?.errors[0]).toContain("required");
  });

  it("says nothing about an optional field that was not found", () => {
    expect(verdict("one field matched and one not")?.valid).toBe(true);
  });

  it("anchors a pattern at the start and not at the end", () => {
    // The reference matches rather than fullmatches, so a rule saying "^A"
    // means "begins with A" and not "is exactly A".
    expect(verdict("a string against a pattern")?.valid).toBe(true);
    expect(verdict("a string failing its pattern")?.valid).toBe(false);
    expect(verdict("a pattern anchored only at the start")?.valid).toBe(true);
  });

  it("prints an allowed set the way the reference does", () => {
    // The message is compared against the reference's own output, and Python
    // renders a list with single quotes around its strings, bare numbers, and
    // a space after each comma. JSON does none of that.
    expect(verdict("allowed values that are numbers")?.errors[0]).toBe(
      "sector: value '42' not in allowed values [1, 2, '3']",
    );
  });

  it("checks a value against its allowed set", () => {
    expect(verdict("a string in its allowed values")?.valid).toBe(true);
    expect(verdict("a string outside its allowed values")?.valid).toBe(false);
  });

  it("validates a float the same way as an int", () => {
    expect(verdict("a float validated as one")?.valid).toBe(true);
  });

  it("accepts a float that happens to be a whole number", () => {
    // The reference converted it with float(), so 2 is a float there however
    // it prints. Demanding a fractional part would reject every round value a
    // screen shows.
    expect(verdict("a float that is a whole number")?.valid).toBe(true);
    expect(extractCase("a float that is a whole number")?.result?.ratio).toBe(2);
  });

  it("reports a value checked against a type it is not", () => {
    // Two configs may name one field: the last to extract wins the value, and
    // every config still validates against it. So a string can be checked as
    // an int, and the message has to say which it found.
    expect(verdict("a value checked against a type it is not")?.errors[0]).toBe("sector: expected int, got str");
    expect(verdict("a fraction checked as a whole number")?.errors[0]).toBe("ratio: expected int, got float");
  });

  it("names the type it found in a mismatch", () => {
    expect(verdict("a whole number checked as text")?.errors[0]).toBe("sector: expected string, got int");
    expect(verdict("a boolean checked as text")?.errors[0]).toBe("docked: expected string, got bool");
    expect(verdict("a boolean checked as a fraction")?.errors[0]).toBe("docked: expected float, got bool");
  });

  it("counts a boolean as a whole number but not as a fraction", () => {
    // A bool is an int in the reference's type system and is not a float.
    // Two configs may name one field, so a value converted as a bool can be
    // validated as a number — refusing it would report an error the
    // reference does not.
    expect(verdict("a boolean checked as a whole number")?.valid).toBe(true);
    expect(verdict("a boolean checked as a fraction")?.valid).toBe(false);
  });

  it("reads a required flag by truthiness", () => {
    // A rules file writing 1 means required. Reading it as optional would let
    // a missing field pass in silence.
    expect(verdict("a required flag that is not a boolean")?.valid).toBe(false);
    expect(verdict("a required flag that is not a boolean")?.errors[0]).toContain("required");
  });

  it("anchors a pattern with no anchor of its own", () => {
    // The reference matches from the start, so a rule saying "lice" does not
    // match "Alice" — it is not a search.
    expect(verdict("a pattern with no anchor")?.valid).toBe(false);
  });

  it("needs no rules to pass", () => {
    expect(verdict("a bool needs no validation rules")?.valid).toBe(true);
  });

  it("reports every problem, not just the first", () => {
    // An operator fixing a rules file wants the whole list rather than one
    // error per run.
    expect(verdict("several problems at once")?.errors).toHaveLength(3);
  });

  it("skips a configuration whose field is not a string", () => {
    // The extraction pass does not check, so the value is present; the
    // validation pass does, so it is never checked. It is the one way a field
    // reaches a caller unvalidated.
    const result = extractCase("a config whose field is not a string")?.result;
    expect(result?.["7"]).toBe("42");
    expect((must(result, "the non-string-field extraction")._validation as { valid: boolean }).valid).toBe(true);
  });
});
